File: rust/src/types/key_value.rs

trait KeyValueStoreBlob {
	fn Get(key: string) -> Result<Vec<u8>, Error>;
	fn Set(key: string, value: Vec<u8>) -> Option<Error>;
	fn Del(key: string) -> Result<Vec<u8>, Error>;
}

trait KeyValueStore<T> {
	fn Get(key: string) -> Result<T, error>;
	fn Set(key: string, value: T) -> Option<Error>;
	fn Del(key: string) -> Result<T, Error>;
}

// type KeyNotFound struct {
// 	Key string
// }

// func (f KeyNotFound) Error() string {
// 	return fmt.Sprintf("key not found: %s", f.Key)
// }

File: go/types/key_value.go
package types

import "fmt"

type KeyValueStoreBlob interface {
	Get(key string) ([]byte, error)
	Set(key string, value []byte) error
	Del(key string) ([]byte, error)
}

type KeyValueStoreInterface interface {
	Get(key string) (interface{}, error)
	Set(key string, value interface{}) error
	Del(key string) (interface{}, error)
}

type KeyValueStore[T any] interface {
	Get(key string) (*T, error)
	Set(key string, value *T) error
	Del(key string) (*T, error)
}

type KeyNotFound struct {
	Key string
}

func (f KeyNotFound) Error() string {
	return fmt.Sprintf("key not found: %s", f.Key)
}
